File: src/discovery.ts
// Resource discovery via `kubectl api-resources`. Lets digg browse *any* kind
// the cluster exposes — built-ins, CRDs, RBAC, HPAs — not just the curated set
// in format.ts. The curated kinds keep their rich columns + detail pages; every
// other discovered kind falls back to a generic list + yaml/describe/edit/events.

interface RunResult {
    stdout: string;
    stderr: string;
    code: number;
}

async function run(args: string[], context?: string): Promise<RunResult> {
    const full = context ? ["--context", context, ...args] : args;
    const proc = Bun.spawn(["kubectl", ...full], { stdout: "pipe", stderr: "pipe" });
    const [stdout, stderr, code] = await Promise.all([
        new Response(proc.stdout).text(),
        new Response(proc.stderr).text(),
        proc.exited,
    ]);
    return { stdout, stderr, code };
}

export interface DiscoveredResource {
    /** Plural resource name used in `kubectl get <name>` (e.g. "horizontalpodautoscalers"). */
    name: string;
    /** PascalCase singular Kind (e.g. "HorizontalPodAutoscaler") — used for events. */
    kind: string;
    /** apiVersion group/version (e.g. "autoscaling/v2"). */
    apiVersion: string;
    namespaced: boolean;
    shortNames: string[];
}

/**
 * Parse `kubectl api-resources -o wide`. Columns are whitespace-aligned:
 *   NAME  SHORTNAMES  APIVERSION  NAMESPACED  KIND  VERBS  CATEGORIES
 * We locate columns by the header's start offsets so values with internal
 * spaces (VERBS like "[get list ...]") don't shift parsing.
 */
const cache = new Map<string, { at: number; value: DiscoveredResource[] }>();
/**
 * Discovery answers nearly every request (a list needs it, a detail page needs
 * it, an events query needs it) and it was a kubectl spawn each time. kubectl
 * itself caches the answer on disk for ten minutes (`--cached`); holding it in
 * memory for five is the same bargain without the process.
 */
const TTL_MS = 5 * 60_000;

export async function apiResources(context: string, opts: { refresh?: boolean } = {}): Promise<DiscoveredResource[]> {
    const hit = cache.get(context);
    if (!opts.refresh && hit && Date.now() - hit.at < TTL_MS) return hit.value;
    const result = await run(["api-resources", "-o", "wide", "--cached", "--verbs=get"], context);
    if (result.code !== 0) {
        // Don't cache a failure: the cluster may just have been unreachable.
        return hit?.value ?? [];
    }
    const value = parseApiResources(result.stdout);
    cache.set(context, { at: Date.now(), value });
    return value;
}

/** Coordinates for a plural resource name, for building its REST path. */
export async function findResource(context: string, kind: string): Promise<DiscoveredResource | undefined> {
    const all = await apiResources(context);
    const lower = kind.toLowerCase();
    return (
        all.find((r) => r.name === kind) ??
        all.find((r) => r.shortNames.includes(kind)) ??
        all.find((r) => r.kind.toLowerCase() === lower) ??
        all.find((r) => r.name === `${lower}s`)
    );
}

export function clearDiscoveryCache(): void {
    cache.clear();
}

/** Pure parser for `kubectl api-resources -o wide` output. Exported for tests. */
export function parseApiResources(stdout: string): DiscoveredResource[] {
    const lines = stdout.split("\n").filter((l) => l.length > 0);
    if (lines.length < 2) {
        return [];
    }
    const header = lines[0];
    const at = (label: string): number => header.indexOf(label);
    const col = { name: at("NAME"), short: at("SHORTNAMES"), api: at("APIVERSION"), ns: at("NAMESPACED"), kind: at("KIND") };
    const out: DiscoveredResource[] = [];
    const seen = new Set<string>();
    for (const line of lines.slice(1)) {
        const slice = (from: number, to: number): string => line.slice(from, to < 0 ? undefined : to).trim();
        const name = slice(col.name, col.short);
        if (!name || seen.has(name)) {
            continue;
        }
        seen.add(name);
        const shortRaw = slice(col.short, col.api);
        out.push({
            name,
            shortNames: shortRaw ? shortRaw.split(",").map((s) => s.trim()).filter(Boolean) : [],
            apiVersion: slice(col.api, col.ns),
            namespaced: slice(col.ns, col.kind).toLowerCase() === "true",
            kind: slice(col.kind, -1).split(/\s+/)[0] ?? "",
        });
    }
    return out.sort((a, b) => a.name.localeCompare(b.name));
}

File: src/internal/kube/discovery.go
package kube

import (
	"sort"
	"strings"
	"sync"
	"time"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

// Resource discovery. Port of src/discovery.ts.
//
// The Bun build ran `kubectl api-resources -o wide` and parsed the output by
// locating the header's start offsets, because values with internal spaces
// (VERBS like "[get list ...]") shift a naive split. The API returns this as
// data, so the parser and its offset arithmetic are gone.
//
// Discovery answers nearly every request (a list needs it, a detail page needs
// it, an events query needs it) and it was a kubectl spawn each time. The
// five-minute memo is kept: kubectl itself caches the same answer on disk for
// ten.

// DiscoveredResource is one kind the cluster exposes.
type DiscoveredResource struct {
	// Name is the plural resource name (e.g. "horizontalpodautoscalers").
	Name string `json:"name"`
	// Kind is the PascalCase singular Kind (e.g. "HorizontalPodAutoscaler"),
	// used for event field selectors.
	Kind string `json:"kind"`
	// APIVersion is the group/version (e.g. "autoscaling/v2").
	APIVersion string   `json:"apiVersion"`
	Namespaced bool     `json:"namespaced"`
	ShortNames []string `json:"shortNames"`
}

const discoveryTTL = 5 * time.Minute

type discoveryCache struct {
	at    time.Time
	value []DiscoveredResource
}

var (
	discoveryMu sync.Mutex
	discovered  = map[string]discoveryCache{}
)

// APIResources lists every kind the cluster exposes that we could list.
func (c *Cluster) APIResources(refresh bool) []DiscoveredResource {
	discoveryMu.Lock()
	hit, ok := discovered[c.Context]
	discoveryMu.Unlock()
	if !refresh && ok && time.Since(hit.at) < discoveryTTL {
		return hit.value
	}

	if refresh {
		c.Discovery.Invalidate()
	}
	// Partial failures are normal: an aggregated API whose backend is down
	// makes ServerPreferredResources return both a list and an error. The list
	// is still worth having — that is exactly the case where hiding every kind
	// would be the wrong answer.
	lists, err := c.Discovery.ServerPreferredResources()
	if err != nil && len(lists) == 0 {
		// Do not cache a failure: the cluster may just have been unreachable.
		return hit.value
	}

	out := []DiscoveredResource{}
	seen := map[string]bool{}
	for _, list := range lists {
		if list == nil {
			continue
		}
		for _, r := range list.APIResources {
			// Subresources (pods/log, deployments/scale) are not listable kinds.
			if strings.Contains(r.Name, "/") {
				continue
			}
			if !verbSupported(r.Verbs, "get") || !verbSupported(r.Verbs, "list") {
				continue
			}
			if seen[r.Name] {
				continue
			}
			seen[r.Name] = true
			out = append(out, DiscoveredResource{
				Name:       r.Name,
				Kind:       r.Kind,
				APIVersion: list.GroupVersion,
				Namespaced: r.Namespaced,
				ShortNames: append([]string{}, r.ShortNames...),
			})
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Name < out[j].Name })

	discoveryMu.Lock()
	discovered[c.Context] = discoveryCache{at: time.Now(), value: out}
	discoveryMu.Unlock()
	return out
}

func verbSupported(verbs metav1.Verbs, want string) bool {
	for _, v := range verbs {
		if v == want {
			return true
		}
	}
	return false
}

// FindResource resolves a kind name against discovery.
func (c *Cluster) FindResource(name string) (DiscoveredResource, bool) {
	for _, d := range c.APIResources(false) {
		if d.Name == name || d.Kind == name {
			return d, true
		}
	}
	return DiscoveredResource{}, false
}
